[CORE-13385,CORE-13982] Expand system iceberg struct with some additional record metadata

oleiman opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR adds `timestamp_type` to the redpanda system struct in iceberg/datalake. This is a non-breaking change and should be processed as a legal schema evolution after upgrade. 

This PR also changes the type of header keys from `binary` to `string`, also in the redpanda struct. This is a table breaking change that we plan to account for by manual migration of customer tables, so this PR also adjusts the datalake upgrade integration test to stop at v25.2.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

### Improvements

* Adds `timestamp_type:integer` to the `redpanda` system iceberg struct.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman committed:
datalake: Add timestamp_type to nested system struct

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
dt/dl: Refactor e2e_test a bit to isolate system type

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
/ci-repeat 1
debug

oleiman commented:
/ci-repeat 1
debug

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73576-->
<!--4dc30d0d-9c30-4d94-aaa2-4011bd69c788-->
<!--641be609-def6-4599-88d2-74d4582c14a3-->
<!--48491797-19c0-4750-b141-ae40fc34f25e-->
<!--23fcb309-fb65-4d81-83ba-56c70f384b65-->
<!--f36e8a11-f3a9-4728-817b-6a5a9a592b19-->
<!--1aa8999a-da87-41d8-b9e5-146fffa8b28d-->
<!--3729b9d0-d942-4af3-bb62-91a1f728685b-->
<!--title-->
## Retry command for [Build#73576](https://buildkite.com/redpanda/redpanda/builds/73576)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/datalake_dlq_test.py::DatalakeDLQTest.test_dlq_table_for_invalid_records@{"catalog_type":"nessie","cloud_storage_type":1,"query_engine":"trino"}
tests/rptest/tests/datalake/datalake_dlq_test.py::DatalakeDLQTest.test_dlq_table_for_invalid_records@{"catalog_type":"nessie","cloud_storage_type":1,"query_engine":"spark"}
tests/rptest/tests/datalake/datalake_dlq_test.py::DatalakeDLQTest.test_dlq_table_for_invalid_records@{"catalog_type":"rest_hadoop","cloud_storage_type":1,"query_engine":"trino"}
tests/rptest/tests/datalake/datalake_dlq_test.py::DatalakeDLQTest.test_dlq_table_for_invalid_records@{"catalog_type":"rest_hadoop","cloud_storage_type":1,"query_engine":"spark"}
tests/rptest/tests/datalake/datalake_dlq_test.py::DatalakeDLQTest.test_dlq_table_for_invalid_records@{"catalog_type":"rest_jdbc","cloud_storage_type":1,"query_engine":"trino"}
tests/rptest/tests/datalake/datalake_dlq_test.py::DatalakeDLQTest.test_dlq_table_for_invalid_records@{"catalog_type":"rest_jdbc","cloud_storage_type":1,"query_engine":"spark"}
tests/rptest/tests/datalake/datalake_verifier_test.py::DatalakeVerifierTest.test_detecting_duplicates@{"cloud_storage_type":1}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--ae41a20a-04aa-46e6-be08-66f1ec48bc17-->
<!--TEST RESULTS-->
<!--e78fb8c9-a77e-42ba-b8a3-2795408101df-->
<!--TEST RESULTS-->
<!--df70ffa8-79ac-47e8-b880-727f3186bb6b-->
<!--TEST RESULTS-->
<!--552a8690-8a90-4006-9164-77c35b4550cb-->
<!--TEST RESULTS-->
<!--20c3c7aa-bff1-474d-8644-07ff85db684e-->
<!--TEST RESULTS-->
<!--a3d42216-2ab6-4dc3-b8db-f2ea353140df-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73576">build#73576</a></summary>


| test_class           | test_method                        | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 | test_history                                                                                                                                                     |
|:---------------------|:-----------------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeDLQTest      | test_dlq_table_for_invalid_records | {"catalog_type": "nessie", "cloud_storage_type": 1, "query_engine": "spark"}      | integration | https://buildkite.com/redpanda/redpanda/builds/73576#0199abf1-0239-4979-94b0-164f2d2bf02b | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDLQTest&test_method=test_dlq_table_for_invalid_records |
| DatalakeDLQTest      | test_dlq_table_for_invalid_records | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "spark"} | integration | https://buildkite.com/redpanda/redpanda/builds/73576#0199abf1-023a-485c-aba8-08d7627657a4 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDLQTest&test_method=test_dlq_table_for_invalid_records |
| DatalakeDLQTest      | test_dlq_table_for_invalid_records | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1, "query_engine": "spark"}   | integration | https://buildkite.com/redpanda/redpanda/builds/73576#0199abf1-023c-4283-8211-6c1720ac5a17 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDLQTest&test_method=test_dlq_table_for_invalid_records |
| DatalakeDLQTest      | test_dlq_table_for_invalid_records | {"catalog_type": "nessie", "cloud_storage_type": 1, "query_engine": "trino"}      | integration | https://buildkite.com/redpanda/redpanda/builds/73576#0199abf1-023e-40f5-83cf-a63ae9946ddd | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDLQTest&test_method=test_dlq_table_for_invalid_records |
| DatalakeDLQTest      | test_dlq_table_for_invalid_records | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"} | integration | https://buildkite.com/redpanda/redpanda/builds/73576#0199abf1-022a-4d89-87ea-17f60592f58b | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDLQTest&test_method=test_dlq_table_for_invalid_records |
| DatalakeDLQTest      | test_dlq_table_for_invalid_records | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1, "query_engine": "trino"}   | integration | https://buildkite.com/redpanda/redpanda/builds/73576#0199abf1-022c-4065-8c8a-1b6eff62e350 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDLQTest&test_method=test_dlq_table_for_invalid_records |
| DatalakeVerifierTest | test_detecting_duplicates          | {"cloud_storage_type": 1}                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/73576#0199abf1-0232-43ef-ad46-79611997308a | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeVerifierTest&test_method=test_detecting_duplicates     |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73842">build#73842</a></summary>


| test_class            | test_method               | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                 |
|:----------------------|:--------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest | test_mount_inexistent     | null                      | integration | https://buildkite.com/redpanda/redpanda/builds/73842#0199c7bf-3d12-402c-8c22-33a44e0f8839 | FLAKY         | 20/21    | upstream reliability is '99.58677685950413'. current run reliability is '95.23809523809523'. drift is 4.34868 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_mount_inexistent    |
| DatalakeVerifierTest  | test_detecting_duplicates | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/73842#0199c7bf-3d12-402c-8c22-33a44e0f8839 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeVerifierTest&test_method=test_detecting_duplicates |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73969">build#73969</a></summary>


| test_class              | test_method                                  | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                             |
|:------------------------|:---------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTest | test_delete_records                          | null             | integration | https://buildkite.com/redpanda/redpanda/builds/73969#0199cf27-406e-4a54-9516-6570ed4eb263 | FLAKY         | 19/21    | upstream reliability is '99.71830985915493'. current run reliability is '90.47619047619048'. drift is 9.24212 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTest&test_method=test_delete_records                |
| DataMigrationsApiTest   | test_creating_and_listing_migrations         | null             | integration | https://buildkite.com/redpanda/redpanda/builds/73969#0199cf1d-078c-4e57-a110-f735eda72a69 | FLAKY         | 20/21    | upstream reliability is '99.7920997920998'. current run reliability is '95.23809523809523'. drift is 4.554 and the allowed drift is set to 50. The test should PASS     | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations |
| SegmentMsTest           | test_segment_rolling_with_retention_consumer | null             | integration | https://buildkite.com/redpanda/redpanda/builds/73969#0199cf27-406d-4bdd-b9ce-77601d61a450 | FLAKY         | 19/21    | upstream reliability is '92.05834683954619'. current run reliability is '90.47619047619048'. drift is 1.58216 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
| TimeQueryTest           | test_timequery_with_local_gc                 | null             | integration | https://buildkite.com/redpanda/redpanda/builds/73969#0199cf27-4068-43a8-817d-73f388ae33b7 | FLAKY         | 14/21    | upstream reliability is '90.13953488372093'. current run reliability is '66.66666666666666'. drift is 23.47287 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TimeQueryTest&test_method=test_timequery_with_local_gc                 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74091">build#74091</a></summary>


| test_class               | test_method                     | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                       |
|:-------------------------|:--------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests     | test_rapid_shadow_link_toggling | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74091#0199def0-7989-4e64-94d6-dc6553c6aaed | FLAKY         | 18/21    | upstream reliability is '88.9908256880734'. current run reliability is '85.71428571428571'. drift is 3.27654 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_rapid_shadow_link_toggling |
| DatalakeUpgradeTest      | test_upload_through_upgrade     | {"cloud_storage_type": 1, "query_engine": "spark"}                                                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/74091#0199def0-7993-4abe-a6dd-b0e740cbfeb7 | FLAKY         | 14/21    | upstream reliability is '100.0'. current run reliability is '66.66666666666666'. drift is 33.33333 and the allowed drift is set to 50. The test should PASS            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeUpgradeTest&test_method=test_upload_through_upgrade      |
| RandomNodeOperationsTest | test_node_operations            | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}        | integration | https://buildkite.com/redpanda/redpanda/builds/74091#0199def0-7989-4e64-94d6-dc6553c6aaed | FLAKY         | 19/21    | upstream reliability is '99.52380952380952'. current run reliability is '90.47619047619048'. drift is 9.04762 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations        |
| RandomNodeOperationsTest | test_node_operations            | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74091#0199def0-798a-43e3-bad0-fdbfc9d8b168 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74221">build#74221</a></summary>


| test_class          | test_method                 | test_arguments                                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 | test_history                                                                                                                                                  |
|:--------------------|:----------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeUpgradeTest | test_upload_through_upgrade | {"cloud_storage_type": 1, "query_engine": "spark"}                                                                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/74221#0199e546-7122-40e8-aa32-6bd52a4f808d | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeUpgradeTest&test_method=test_upload_through_upgrade |
| TieredStorageTest   | test_tiered_storage         | {"cloud_storage_type_and_url_style": [1, "virtual_host"], "test_case": {"name": "(TS_Read == True, TS_Timequery == True, SpilloverManifestUploaded == True)"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74221#0199e546-712b-44e4-86d6-28d0ef594622 | FAIL          | 0/1      |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TieredStorageTest&test_method=test_tiered_storage           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74285">build#74285</a></summary>


| test_class                      | test_method                          | test_arguments                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                             |
|:--------------------------------|:-------------------------------------|:--------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest           | test_creating_and_listing_migrations | null                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74285#0199e9ac-fe75-41ab-8a7c-09a2732f98c8 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test             | {"cloud_storage_type": 1, "mixed_versions": true} | integration | https://buildkite.com/redpanda/redpanda/builds/74285#0199e9a6-9de9-4cc3-897b-30001cb8cb99 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test   |
| AdminSecurityReportTest         | test_security_report                 | null                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74285#0199e9a6-9de1-4ba0-a8c4-4bd2ef47d5c7 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AdminSecurityReportTest&test_method=test_security_report               |
| TimeQueryTest                   | test_timequery_with_local_gc         | null                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74285#0199e9ac-fe79-439c-8af9-fd11a8f8f6bf | FLAKY         | 19/21    | upstream reliability is '90.11473962930273'. current run reliability is '90.47619047619048'. drift is -0.36145 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TimeQueryTest&test_method=test_timequery_with_local_gc                 |
| RapidTopicRecreateTest          | test_topic_rapid_recreation          | null                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74285#0199e9a6-9ddd-43cb-8dde-6105b7643a7f | FLAKY         | 20/21    | upstream reliability is '98.9406779661017'. current run reliability is '95.23809523809523'. drift is 3.70258 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RapidTopicRecreateTest&test_method=test_topic_rapid_recreation         |
</details>
<!--/body-->

oleiman commented:
/ci-repeat 1
debug

oleiman commented on src/v/datalake/table_definition.cc:
```diff
@@ -38,6 +38,9 @@ struct_type schemaless_struct_type() {
       nested_field::create(10, "topic", field_required::no, string_type{}));
+    system_fields.fields.emplace_back(
+      nested_field::create(
+        11, "timestamp_type", field_required::no, string_type{}));
```
oops, should have been just the int

oleiman committed:
model: Fix timestemp_type values

We'll be writing these out to parquet, so it's good to be explicit.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
/ci-repeat 1
debug

oleiman commented:
/ci-repeat 1
debug

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73842-->
<!--3e99a03d-ea2d-4e4d-9bc4-5598dc7561b3-->
<!--title-->
## Retry command for [Build#73842](https://buildkite.com/redpanda/redpanda/builds/73842)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/datalake_verifier_test.py::DatalakeVerifierTest.test_detecting_duplicates@{"cloud_storage_type":1}
```
<!--/body-->

rockwotj reviewed (commented):
Generally am good with this change although I expect basically nobody to actually care about this information outside of a select few. Should we wait to merge this until we have dictionary encoding? Right now we will have to write a 0 or 1 for every row in the parquet file. It should compress well however.

oleiman commented:
> Generally am good with this change although I expect basically nobody to actually care about this information outside of a select few. Should we wait to merge this until we have dictionary encoding? Right now we will have to write a 0 or 1 for every row in the parquet file. It should compress well however.

@rockwotj fine by me 🤷. my last sync point w/ @mattschumpert concluded that we would add this for 25.3, but I have no objections if the winds are shifting. I just want to make sure this is ready to go if and when we decide to merge it.

rockwotj commented on src/v/datalake/record_translator.cc:
```diff
@@ -150,17 +156,25 @@ default_translator::translate_data(
     co_return co_await kv_translator.translate_data(
-      pid, o, std::move(key), val_type, std::move(parsable_val), ts, headers);
+      pid,
+      o,
+      std::move(key),
+      val_type,
+      std::move(parsable_val),
+      ts,
+      ts_t,
+      headers);
 }
 
 record_type key_value_translator::build_type(std::optional<resolved_type>) {
     auto ret_type = schemaless_struct_type();
     ret_type.fields.emplace_back(
       iceberg::nested_field::create(
-        10, "value", iceberg::field_required::no, iceberg::binary_type{}));
+        11, "value", iceberg::field_required::no, iceberg::binary_type{}));
```
Sorry I am a bit naive here, is changing this field ID here an issue or do we overwrite it later?

oleiman replied:
hrm, good point. i think these might be placeholders but I will confirm.

oleiman replied:
> placeholders

looking back over the code, yeah i'm fairly sure the IDs we fill here are totally meaningless. even if the table is brand new, we should generate IDs further downstream. @nvartolomei maybe can confirm that's still the case

andrwng replied:
Yeah, these are placeholders, actual field assignments will be done by the catalog

oleiman replied:
ace. ty

rockwotj reviewed (dismissed):
LGTM, as discussed I think we can remove the feature flag since there will be a few things here

oleiman committed:
datalake: Change header keys from binary -> string

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
dt/dl: Fix databricks test to use new redpanda field schema

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
@rockwotj @nvartolomei - can someone remind me of the rationale for ditching the feature flag? are we just going to require disabling iceberg before upgrade?

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74221-->
<!--84bdfb21-5dc0-418c-be9a-2d7553300093-->
<!--title-->
## Retry command for [Build#74221](https://buildkite.com/redpanda/redpanda/builds/74221)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/datalake_upgrade_test.py::DatalakeUpgradeTest.test_upload_through_upgrade@{"cloud_storage_type":1,"query_engine":"spark"}
```
<!--/body-->

oleiman committed:
dt/rptest: Add optional max_version to load_version_range

By default returns [initial, HEAD] as before. If max_version is supplied,
return [initial, max] instead.

The result is an inclusive version range in either case.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

andrwng commented on src/v/datalake/table_definition.cc:
```diff
@@ -36,6 +36,9 @@ struct_type schemaless_struct_type() {
 
     system_fields.fields.emplace_back(
       nested_field::create(9, "key", field_required::no, binary_type{}));
```
Just want to double check, is redpanda.key still supposed to be a binary field?

oleiman replied:
yeah afaik. all i heard was that we were breaking headers

rockwotj replied:
yes

rockwotj replied:
people put avro and other stuff in record keys. It's just header keys that are different
